src/webgpu/shader/validation/expression/call/builtin/subgroupBallot.spec.ts

```diff
@@ -0,0 +1,154 @@
+
+const kArgumentTypes = objectsToRecord(kAllScalarsAndVectors);
+
+// Should this be in the spec as an alternative to subgroupBallot(true)?
+g.test('no_params')
+  .desc('Validates no parameter variant')
+  .params(u =>
+    u.combine('type', keysOf(kArgumentTypes)).filter(t => {
+      const type = kArgumentTypes[t.type];
+      const eleType = elementTypeOf(type);
+      return eleType !== Type.abstractInt && eleType !== Type.abstractFloat;
+    })
+  )
+  .beforeAllSubcases(t => {
+    const features = ['subgroups' as GPUFeatureName];
+    const type = kArgumentTypes[t.params.type];
+    if (type.requiresF16()) {
+      features.push('subgroups-f16' as GPUFeatureName);
+      features.push('shader-f16');
+    }
+    t.selectDeviceOrSkipTestCase(features);
+  })
+  .fn(t => {
+    const type = kArgumentTypes[t.params.type];
+    let enables = `enable subgroups;\n`;
+    if (type.requiresF16()) {
+      enables += `enable subgroups_f16;\nenable f16;`;
+    }
+    const wgsl = `
+${enables}
+@compute @workgroup_size(1)
+fn main() {
+  let res : ${type.toString()} = subgroupBallot();
+}`;
+
+    t.expectCompileResult(type === Type.vec4u, wgsl);
+  });
+
+g.test('data_type')
+  .desc('Validates data parameter type')
+  .params(u => u.combine('type', keysOf(kArgumentTypes)))
+  .beforeAllSubcases(t => {
+    const features = ['subgroups' as GPUFeatureName];
+    const type = kArgumentTypes[t.params.type];
+    if (type.requiresF16()) {
+      features.push('subgroups-f16' as GPUFeatureName);
+      features.push('shader-f16');
+    }
+    t.selectDeviceOrSkipTestCase(features);
+  })
+  .fn(t => {
+    const type = kArgumentTypes[t.params.type];
+    let enables = `enable subgroups;\n`;
+    if (type.requiresF16()) {
+      enables += `enable subgroups_f16;\nenable f16;`;
+    }
+    const wgsl = `
+${enables}
+@compute @workgroup_size(1)
+fn main() {
+  _ = subgroupBallot(${type.create(0).wgsl()}, 0);
+}`;
+
+    t.expectCompileResult(type === Type.bool, wgsl);
+  });
+
+g.test('return_type')
+  .desc('Validates data parameter type')
+  .params(u =>
+    u.combine('type', keysOf(kArgumentTypes)).filter(t => {
+      const type = kArgumentTypes[t.type];
+      const eleType = elementTypeOf(type);
+      return eleType !== Type.abstractInt && eleType !== Type.abstractFloat;
+    })
+  )
+  .beforeAllSubcases(t => {
+    const features = ['subgroups' as GPUFeatureName];
+    const type = kArgumentTypes[t.params.type];
+    if (type.requiresF16()) {
+      features.push('subgroups-f16' as GPUFeatureName);
+      features.push('shader-f16');
+    }
+    t.selectDeviceOrSkipTestCase(features);
+  })
+  .fn(t => {
+    const type = kArgumentTypes[t.params.type];
+    let enables = `enable subgroups;\n`;
+    if (type.requiresF16()) {
+      enables += `enable subgroups_f16;\nenable f16;`;
+    }
+    const wgsl = `
+${enables}
+@compute @workgroup_size(1)
+fn main() {
+  let res : ${type.toString()} = subgroupBallot(${Type.bool.create(0).wgsl()});
```

Comment:
The arg for subgroupBallot will just be `false` here.  May as well type it in directly?